src/libexpr/primops/regex.cc

```diff
@@ -0,0 +1,360 @@
+}
+
+size_t regexCacheSize(std::shared_ptr<RegexCache> cache)
+{
+    return cache->keys.size();
+}
+
+static void extract_matches(EvalState & state, const PCRE::MatchData& match, Value & v)
+{
+    auto nameTable = match.regex().nameTable();
+    if (nameTable.size()) {
+        // try to extract named bindings
+        auto bindings = state.buildBindings(nameTable.size());
+        for (auto i : nameTable) {
+            Value & elem = bindings.alloc(i.first);
+            if (!match[i.second].has_value()) elem.mkNull();
+            else elem.mkString(*match[i.second]);
+        }
+        v.mkAttrs(bindings);
+    } else {
+        // the first match is the whole string
+        const size_t len = match.size() - 1;
+        state.mkList(v, len);
+        for (size_t i = 0; i < len; ++i) {
+            if (!match[i+1].has_value())
+                (v.listElems()[i] = state.allocValue())->mkNull();
+            else
+                (v.listElems()[i] = state.allocValue())->mkString(*match[i + 1]);
+        }
+    }
+}
+
+void prim_match(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
+
+    try {
+
+        auto& regex = state.regexCache->get(re);
+
+        PathSet context;
+        const auto str = state.forceString(*args[1], context, pos);
+
+        PCRE::MatchData match(regex);
+        int rc = regex.match(str, match, 0, PCRE2_ANCHORED | PCRE2_ENDANCHORED);
+        if (rc == PCRE2_ERROR_NOMATCH) {
+            v.mkNull();
+            return;
+        }
+        extract_matches(state, match, v);
+
+    } catch (RegexError & e) {
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("error while evaluating regex '%s': ", re, e.what()),
+            .errPos = state.positions[pos]
+        }));
+    }
+}
+
+static RegisterPrimOp primop_match({
+    .name = "__match",
+    .args = {"regex", "str"},
+    .doc = R"s(
+      Returns a list if the [extended POSIX regular
+      expression](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04)
+      *regex* matches *str* precisely, otherwise returns `null`. Each item
+      in the list is a regex group.
+
+      ```nix
+      builtins.match "ab" "abc"
+      ```
+
+      Evaluates to `null`.
+
+      ```nix
+      builtins.match "abc" "abc"
+      ```
+
+      Evaluates to `[ ]`.
+
+      ```nix
+      builtins.match "a(b)(c)" "abc"
+      ```
+
+      Evaluates to `[ "b" "c" ]`.
+
+      ```nix
+      builtins.match "[[:space:]]+([[:upper:]]+)[[:space:]]+" "  FOO   "
+      ```
+
+      Evaluates to `[ "FOO" ]`.
+    )s",
+    .fun = prim_match,
+});
+
+/* Split a string with a regular expression, and return a list of the
+   non-matching parts interleaved by the lists of the matching groups. */
+void prim_split(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
```

Comment:
same here, asserts would improve debuggability if we ever end up with code that inserts nullptrs